dali/ft/filecopy.cpp

```diff
@@ -2596,6 +2639,325 @@ void FileSprayer::pushParts()
+            else if (tstatus==Blobs::Models::CopyStatus::Pending)
+                status = ApiCopyStatus::Pending;
+            else if (tstatus==Blobs::Models::CopyStatus::Aborted)
+                status = ApiCopyStatus::Aborted;
+            else
+                status = ApiCopyStatus::Failed;
+        }
+        catch(const Azure::Core::RequestFailedException& e)
+        {
+            IERRLOG("Transfer using API .Poll() failed: %s (code %d)", e.ReasonPhrase.c_str(), static_cast<int>(e.StatusCode));
+            return ApiCopyStatus::Failed;
+        }
+        return status;
+    }
+    ApiCopyStatus abortCopy()
+    {
+        try
+        {
+            blobCopyOp->Poll();
+            Models::BlobProperties props = blobCopyOp->Value();
+            if (props.CopyStatus.HasValue() && props.CopyStatus.Value()!=Blobs::Models::CopyStatus::Success)
+            {
+                if (blobCopyOp->HasValue() && blobCopyOp->Value().CopyId.HasValue())
+                    blobClient->AbortCopyFromUri(blobCopyOp->Value().CopyId.Value().c_str());
+                else
+                    IERRLOG("Transfer using API .AbortCopyFromUri() failed: unexpected response from API");
+                status = ApiCopyStatus::Aborted;
+            }
+            else
+                status = ApiCopyStatus::Success;
+        }
+        catch(const Azure::Core::RequestFailedException& e)
+        {
+            UWARNLOG("Abort copy operation failed: %s (code %d)", e.ReasonPhrase.c_str(), static_cast<int>(e.StatusCode));
+            status = ApiCopyStatus::Failed;
+        }
+        return status;
+    }
+};
+
+bool FileSprayer::transferUsingAPI()
+{
+    // N.B. Only call this member function, if FileSprayer::useAPICopy() has returned true
+    // Source
+    StringBuffer sourceClusterName, sourceBaseURI, sourceSAStoken;
+    distributedSource->getClusterName(0, sourceClusterName);
+    Owned<IStoragePlane> sourcePlane = getDataStoragePlane(sourceClusterName.str(), false);
+    if (!CApiCopyClient::buildAzureStorageURIBase(sourcePlane, sourceBaseURI))
+        return false;
+    sourcePlane->getSASToken(sourceSAStoken);
+
+    // Target
+    StringBuffer targetClusterName, targetBaseURI, targetSAStoken;
+    distributedTarget->getClusterName(0, targetClusterName);
+    Owned<IStoragePlane> targetPlane = getDataStoragePlane(targetClusterName.str(), false);
+    if (!CApiCopyClient::buildAzureStorageURIBase(targetPlane, targetBaseURI))
+        return false;
+    targetPlane->getSASToken(targetSAStoken);
+    StorageType targetType = targetPlane->getStorageType();
+
+    OwnedPointerArrayOf<IAPICopyClient> apiClients;
+    unsigned failCount = 0;
+    ForEachItemIn(idx1, partition)
+    {
+        PartitionPoint & cur = partition.item(idx1);
+        OutputProgress & curProgress = progress.item(idx1);
+        if (curProgress.status==OutputProgress::StatusCopied)
+        {
+            apiClients.append(nullptr);
```

Comment:
Worth adding a comment to indicate that null values are used to represent parts that have already been copied.